File: ChipTallyiOS/ChipTally/Domain/Entities/Transaction.swift
//
//  Transaction.swift
//  ChipTally
//

import Foundation

enum TransactionType: String, Codable {
    /// 플레이어끼리 1:1 로 주고받음.
    case transfer
    /// 플레이어가 팟에 걸었음. `toPlayerId` 는 팟 자리표시자다.
    case bet
    /// 플레이어가 팟을 전부 가져감. `fromPlayerId` 는 팟 자리표시자다.
    case potWin
}

struct Transaction: Identifiable {
    let id: UUID
    let fromPlayerId: UUID
    let toPlayerId: UUID
    let amount: Int
    let timestamp: Date
    let type: TransactionType

    init(
        id: UUID = UUID(),
        fromPlayerId: UUID,
        toPlayerId: UUID,
        amount: Int,
        timestamp: Date = Date(),
        type: TransactionType = .transfer
    ) {
        self.id = id
        self.fromPlayerId = fromPlayerId
        self.toPlayerId = toPlayerId
        self.amount = amount
        self.timestamp = timestamp
        self.type = type
    }

    static func transfer(fromPlayerId: UUID, toPlayerId: UUID, amount: Int) -> Transaction {
        Transaction(fromPlayerId: fromPlayerId, toPlayerId: toPlayerId, amount: amount, type: .transfer)
    }

    static func bet(playerId: UUID, amount: Int) -> Transaction {
        Transaction(fromPlayerId: playerId, toPlayerId: Transaction.potId, amount: amount, type: .bet)
    }

    static func potWin(playerId: UUID, amount: Int) -> Transaction {
        Transaction(fromPlayerId: Transaction.potId, toPlayerId: playerId, amount: amount, type: .potWin)
    }

    /// 사람이 아닌 '팟' 을 가리키는 자리표시자. 어떤 플레이어 id 와도 겹치지 않는다.
    static let potId = UUID(uuidString: "00000000-0000-0000-0000-000000000000")!
}

File: ChipTallyAOS/app/src/main/java/com/chiptally/domain/model/Transaction.kt
package com.chiptally.domain.model

import java.util.Date
import java.util.UUID

enum class TransactionType {
    /** 플레이어끼리 1:1 로 주고받음. */
    TRANSFER,

    /** 플레이어가 팟에 걸었음. [Transaction.toPlayerId] 는 비어 있다. */
    BET,

    /** 플레이어가 팟을 전부 가져감. [Transaction.fromPlayerId] 는 비어 있다. */
    POT_WIN
}

data class Transaction(
    val id: String = UUID.randomUUID().toString(),
    val fromPlayerId: String,
    val toPlayerId: String,
    val amount: Int,
    val timestamp: Date = Date(),
    // Gson 은 Unsafe 로 객체를 만들어 생성자 기본값을 무시한다. 이 필드가 없던 시절에
    // 저장된 세션을 읽으면 null 이 들어오므로 nullable 로 두고 [type] 으로 읽는다.
    private val transactionType: TransactionType? = null
) {
    val type: TransactionType
        get() = transactionType ?: TransactionType.TRANSFER

    companion object {
        fun transfer(fromPlayerId: String, toPlayerId: String, amount: Int) = Transaction(
            fromPlayerId = fromPlayerId,
            toPlayerId = toPlayerId,
            amount = amount,
            transactionType = TransactionType.TRANSFER
        )

        fun bet(playerId: String, amount: Int) = Transaction(
            fromPlayerId = playerId,
            toPlayerId = "",
            amount = amount,
            transactionType = TransactionType.BET
        )

        fun potWin(playerId: String, amount: Int) = Transaction(
            fromPlayerId = "",
            toPlayerId = playerId,
            amount = amount,
            transactionType = TransactionType.POT_WIN
        )
    }
}
